docs/design/yaml-file.rst

```diff
@@ -0,0 +1,165 @@
+
+Global settings
+~~~~~~~~~~~~~~~
+
+Those settings will be read from the YAML file on the ``default branch``.
+
+- Language
+- Programming Language
+- Project homepage
+- Tags
+- Analytics code
+- Redirects 
+
+Local settings
+~~~~~~~~~~~~~~
+
+Those configurations will be read from the YAML file in the current version that is being built.
+
+Several settings are already implemented and documented on
+https://docs.readthedocs.io/en/latest/yaml-config.html.
+So, they aren't covered with much detail here. 
+
+- Documentation type
+- Project installation (virtual env, requirements file, sphinx configuration file, etc)
+- Additional builds (pdf, epub)
+- Python interpreter
+
+Configuration file
+------------------
+
+Format
+~~~~~~
+
+The file format is based on the YAML spec 1.2 [#yaml-spec]_
+(latest version on the time of this writing).
+
+The file must be on the root directory of the repository, and must be named as:
+
+- ``.readthedocs.yml``
+- ``readthedocs.yml``
+
+Conventions
+~~~~~~~~~~~
+
+The spec of the configuration file must use this conventions.
+
+- Use `[]` to indicate an empty list
+- Use `null` to indicate a null value
+- Use `all` (internal string keyword) to indicate that all options are included on a list with predetermined choices.
+- Use `true`  and `false` as only options on boolean fields
+
+Spec
+~~~~
+
+The current spec is documented on https://docs.readthedocs.io/en/latest/yaml-config.html
+and https://github.com/rtfd/readthedocs-build/blob/master/docs/spec.rst
+
+Versioning the spec
+-------------------
+
+TODO
+
+Adoption of the configuration file
+----------------------------------
+
+When a user creates a new project or it's on the settings page,
+we could suggest her/him an example of a functional configuration file with a minimal setup.
+
+Main source for global configurations
+-------------------------------------
+
+There are some global settings that are needed for the build process (like language).
+So it's needed to read this configurations from one source of truth before each build.
+This source can be taken from the ``default branch`` setting.
+RTD will checkout to this branch and read this configurations before the real build process starts.
+
+That solves one problem, but RTD still need to know when to update the others global settings.
+Would be a waste of resources to made a new build each time a global setting is updated for it to take effect.
+Currently, RTD keeps a dedicated local repository for each version, which is updated before a build.
+RTD could have a central repository for this operations [#one-checkout]_.
+
+Configuration file and database options
+---------------------------------------
+
+To decouple the configuration file from the database and keep the compatibility with projects without one,
+we need to generate a YAML file from the existing database options,
+this will also help with the `Adoption of the configuration file`_.
```

Comment:
After writing my previuos comment and reading this, I think this could be kind of complex: each time a new option is added we will need to tweak our "DB to YAML" builder to handle the new option/format/etc.

Reply:
But this can be done only for previous projects, new ones can be forced to use the yaml file.

Reply:
Mmm... I'm not sure we can force the user to have a yaml file in the repo. I suppose that RTD tries to "just works" even when the repo only has an `index.rst` and nothing else. Forcing to have a yaml will break that idea.

To be considered.

Reply:
Maybe autogenerate a configuration file for new projects like the defaults that are created when there isn't a sphinx project

Reply:
We should also only be converting DB -> YAML for options that exist in the DB. Presumably we won't be adding *new* DB options, so the DB -> YAML conversion should stay the same. If we add new YAML options, unless they are required, the conversion process will just not set them.

Reply:
I don't think we want to "convert" any database state. We don't want to decouple the actual database usage with our config file, just supplement it. Eventually, new projects don't have any version-level config in our database and we can hide the admin for these users. But, removing the database shouldn't be the focus. I think that removes the need to generate a config file, besides a basic suggestion file for old users.